subst: delay replacing node until pass end

kamilrakoczy opened the pull request:
This way we can delete assignment node immediately after replacing rhs values and reduce total ram usage.

This approach allows us to remove restriction on ``SUBST_MAX_OPS_SUBST`` without drastically increase ram usage, but it increases pass time, as we need to iterate more nodes.
Some statistics on our test design (from ``stats.txt`` file). Please note, that because of such high memory usage, I'm using zram swap, so time measurements can be very biased and vary from run to run.
|         | current master | current master without ``SUBST_MAX_OPS_SUBST`` | after this patch  | after this patch keeping ``SUBST_MAX_OPS_SUBST`` |
|---------|----|----|---|---|
| subst pass time |  80.15s | 91.34s | 100.31s  | 95.71s |
| RSS after subst pass |  69 987MB |  87 558MB   | 69 051MB  | 69 062MB |
| increase in RSS from previous pass | 1 060MB | 18 636MB | 126MB | 126MB |

I've also checked, that resulted cpp files are identical, when we compare output from ``after this patch`` and ``current master without SUBST_MAX_OPS_SUBST`` (comparing ``after this patch keeping SUBST_MAX_OPS_SUBST`` to ``current master `` doesn't give identical output as we differently increase ``m_ops`` value) .


Signed-off-by: Kamil Rakoczy <krakoczy@antmicro.com>

kamilrakoczy commented:
As I've noticed, that I'm using zram swap and the measurements could be very biased, I've redid them on the machine that could fit entire design in the RAM:
|         | current master | current master without ``SUBST_MAX_OPS_SUBST`` | after this patch  | after this patch keeping ``SUBST_MAX_OPS_SUBST`` |
|---------|----|----|---|---|
| subst pass time |  33.23s | 96.26s | 83.48s  | 46.73s |
| RSS after subst pass |  70 013MB |  87 577MB   | 69 075MB  | 69 074MB |
| increase in RSS from previous pass | 1 062MB | 18 629MB | 128MB | 128MB |

wsnyder commented:
Can you add total times to the above table - that is does the added 13 seconds to subst just take time that was later in V3Const?  It looks attractive but don't want to be introducing a O(n^2) slowdown or something.

wsnyder commented on src/V3Subst.cpp:
```diff
@@ -156,7 +159,33 @@ class SubstVarEntry final {
+        if (debug() > 5) nodep->dumpTree(cout, " substw_old: ");
+        AstNode* newp = substp;
+        if (!nodep->isQuad() && newp->isQuad()) {
+            newp = new AstCCast(newp->fileline(), newp, nodep);
```
```suggestion
            newp = new AstCCast{newp->fileline(), newp, nodep};
```

wsnyder commented on src/V3Subst.cpp:
```diff
@@ -156,7 +159,33 @@ class SubstVarEntry final {
+        if (!assignp) return statSubsts;
+        for (unsigned i = 0; i < assigns.size(); i++) {
+            VL_DO_DANGLING(replaceSubstEtc(assigns[i], assignp->cloneTree(true)), assigns[i]);
+            statSubsts++;
```
```suggestion
            ++statSubsts;
```

wsnyder commented on src/V3Subst.cpp:
```diff
@@ -156,7 +159,33 @@ class SubstVarEntry final {
+    unsigned replaceAssignments(AstNode* assignp, std::vector<AstNode*>& assigns) {
+        unsigned statSubsts = 0;
+        if (!assignp) return statSubsts;
+        for (unsigned i = 0; i < assigns.size(); i++) {
```
```suggestion
        for (unsigned i = 0; i < assigns.size(); ++i) {
```

wsnyder commented on src/V3Subst.cpp:
```diff
@@ -156,7 +159,33 @@ class SubstVarEntry final {
+        unsigned statSubsts = 0;
+        AstNode* assignp = substWhole(m_whole.m_assignp);
+        statSubsts += replaceAssignments(assignp, m_whole.m_assigns);
+        for (unsigned i = 0; i < m_words.size(); i++) {
```
```suggestion
        for (unsigned i = 0; i < m_words.size(); ++i) {
```

wsnyder commented on src/V3Subst.cpp:
```diff
@@ -229,17 +272,15 @@ class SubstVisitor final : public SubstBaseVisitor {
     // AstVar::user2            -> int step number for last assignment, 0=not set yet
     const VNUser1InUse m_inuser1;
     const VNUser2InUse m_inuser2;
+    const VNUser3InUse m_inuser3;
```
Need to add comment above this section as to how user3 is used.

wsnyder commented on src/V3Subst.cpp:
```diff
@@ -368,11 +400,11 @@ class SubstVisitor final : public SubstBaseVisitor {
-        V3Stats::addStat("Optimizations, Substituted temps", m_statSubsts);
         for (SubstVarEntry* ip : m_entryps) {
-            ip->deleteUnusedAssign();
+            m_statSubsts += ip->deleteUnusedAssign();
```
I'm not seeing the advantage of passing the statistic up as a return value, just increment it directly in the lower function.

kamilrakoczy commented:
I've fixed review comments, rebased PR and remeasured times. I've run test 8 times and took average result:
|                                | current master | after this patch | after this patch keeping   ``SUBST_MAX_OPS_SUBST`` |
|--------------------------------|----------------|------------------|----------------------------------------------------|
| Total execution time (s)  | 1102.82        | 1158.9575        | 1129.52                                            |
| max RSS      (MB)                  | 68198.16     | 63324.17       | 63320.16                                           |

Here is also table with execution times of each stage (this time it is from single run, not average):
|                   | current master | after this patch | after this patch keeping   ``SUBST_MAX_OPS_SUBST`` |
|-------------------|----------------|------------------|----------------------------------------------------|
| 074_const_cpp     | 37.700751      | 38.736732        | 38.793373                                          |
| 075_subst         | 31.691712      | 59.077053        | 47.310752                                          |
| 076_const_cpp     | 20.109523      | 39.381419        | 23.876688                                          |
| 077_deadAll       | 8.098396       | 8.366679         | 8.243275                                           |
| 078_merge_cond    | 10.802252      | 11.113922        | 11.01348                                           |
| 079_reloop        | 1.437555       | 1.386786         | 1.399331                                           |
| 080_depth         | 6.332156       | 6.488136         | 6.372582                                           |
| 081_cast          | 16.643724      | 17.082588        | 16.978002                                          |
| 082_common        | 0.215075       | 0.216401         | 0.238486                                           |
| 083_variableorder | 0.260235       | 0.259015         | 0.3578                                             |
| 084_cuse          | 7.5166         | 7.75594          | 7.971389                                           |
| 085_emit          | 144.293101     | 146.619975       | 149.468844                                         |

wsnyder commented:
Can you post the RSS sizes after the patch - what is the max RSS across all passes vs the size of the RSS with this patch?

kamilrakoczy commented:
I'm not sure if I understood correctly what you meant, here is table comparing max RSS sizes of different passes:
|Stage, Memory (MB)|current master|after this patch|after this patch keeping ``SUBST_MAX_OPS_SUBST``|
|-------------------|--------------|----------------|------------------------------------------------|
|071_clean          |58399.164062  |58399.164062    |58399.171875                                    |
|072_premit         |58399.164062  |58399.164062    |58399.171875                                    |
|073_expand         |63123.164062  |63121.164062    |63124.171875                                    |
|074_const_cpp      |63123.164062  |63121.164062    |63124.171875                                    |
|075_subst          |68198.164062  |63320.164062    |63324.171875                                    |
|076_const_cpp      |68198.164062  |63320.164062    |63324.171875                                    |
|077_deadAll        |68198.164062  |63320.164062    |63324.171875                                    |
|078_merge_cond     |68198.164062  |63320.164062    |63324.171875                                    |
|079_reloop         |68198.164062  |63320.164062    |63324.171875                                    |
|080_depth          |68198.164062  |63320.164062    |63324.171875                                    |
|081_cast           |68198.164062  |63320.164062    |63324.171875                                    |
|082_common         |68198.164062  |63320.164062    |63324.171875                                    |
|083_variableorder  |68198.164062  |63320.164062    |63324.171875                                    |
|084_cuse           |68198.164062  |63320.164062    |63324.171875                                    |
|085_emit           |68198.164062  |63320.164062    |63324.171875                                    |

wsnyder commented:
Thanks, so this patch is reducing the max RSS, that is no later stage is using more memory.

While the memory savings are good, I'm still concerned about accepting the runtime increase.  I suspect that the memory is from the delayed pushDeletep's in the current algorithm and doing those sooner (e.g. calling doDeletes() per module) might give us most of the same memory savings without any runtime increase.... Can/did you experiment with that?

kamilrakoczy commented:
I've tried approach with calling doDeletes per module, it doesn't give as big reduce on max RSS, but it doesn't increase runtime.

I think it is worth merging doDeletes per module, but I will continue working on reducing runtime in my previous approach.

Deleting nodes per module gives max RSS: ``66990.47MB`` and subst time: ``33.04s``
